xds/src/test/java/io/grpc/xds/SharedXdsClientPoolProviderTest.java

```diff
@@ -0,0 +1,247 @@
+      assertThat(xdsClientPools[i]).isSameInstanceAs(xdsClientPools[0]);
+    }
+    verify(bootstrapper).readBootstrap();
+    verify(channelFactory).createChannel(ArgumentMatchers.<ServerInfo>anyList());
+    executor.shutdown();
+  }
+
+  @Test
+  public void refCountedXdsClientObjectPool_delayedCreation() {
+    RefCountedXdsClientObjectPool xdsClientPool =
+        new RefCountedXdsClientObjectPool(channel, node, factory);
+    assertThat(xdsClientRef.get()).isNull();
+    xdsClientPool.getObject();
+    assertThat(xdsClientRef.get()).isNotNull();
+  }
+
+  @Test
+  public void refCountedXdsClientObjectPool_refCounted() {
+    RefCountedXdsClientObjectPool xdsClientPool =
+        new RefCountedXdsClientObjectPool(channel, node, factory);
+
+    // getObject once
+    XdsClient xdsClient = xdsClientPool.getObject();
+    assertThat(xdsClient).isNotNull();
+    // getObject twice
+    assertThat(xdsClientPool.getObject()).isSameInstanceAs(xdsClient);
+    // returnObject once
+    assertThat(xdsClientPool.returnObject(xdsClient)).isNull();
+    verify(xdsClient, never()).shutdown();
+    // returnObject twice
+    assertThat(xdsClientPool.returnObject(xdsClient)).isNull();
+    verify(xdsClient).shutdown();
+  }
+
+  @Test
+  public void refCountedXdsClientObjectPool_getObjectCreatesNewInstanceIfAlreadyShutdown() {
+    RefCountedXdsClientObjectPool xdsClientPool =
+        new RefCountedXdsClientObjectPool(channel, node, factory);
+    XdsClient xdsClient1 = xdsClientPool.getObject();
+    verify(xdsClient1, never()).shutdown();
+    assertThat(xdsClientPool.returnObject(xdsClient1)).isNull();
+    verify(xdsClient1).shutdown();
+
+    XdsClient xdsClient2 = xdsClientPool.getObject();
+    assertThat(xdsClient2).isNotSameInstanceAs(xdsClient1);
+  }
+
+  @Test
+  public void refCountedXdsClientObjectPool_concurrentGet() throws InterruptedException {
+    int numThreads = 3;
+    ExecutorService executor = Executors.newFixedThreadPool(numThreads);
+    final RefCountedXdsClientObjectPool xdsClientPool =
+        new RefCountedXdsClientObjectPool(channel, node, factory);
+    final XdsClient[] xdsClients = new XdsClient[numThreads];
+    final CountDownLatch latch = new CountDownLatch(numThreads);
+    class GetXdsClient implements Runnable {
+      private final int index;
+
+      private GetXdsClient(int index) {
+        this.index = index;
+      }
+
+      @Override
+      public void run() {
+        xdsClients[index] = xdsClientPool.getObject();
+        latch.countDown();
+      }
+    }
+
+    for (int i = 0; i < numThreads; i++) {
+      executor.execute(new GetXdsClient(i));
+    }
+    latch.await(1L, TimeUnit.SECONDS);
+    assertThat(xdsClientRef.get()).isNotNull();
+    for (int i = 0; i < numThreads; i++) {
+      assertThat(xdsClients[i]).isSameInstanceAs(xdsClientRef.get());
+    }
+    executor.shutdown();
+  }
+
+  @Test
+  public void refCountedXdsClientObjectPool_concurrentReturn() throws InterruptedException {
```

Comment:
This kind of test runs non-deterministically, not fix for unit tests. I don't think we have to cover concurrency in unit tests.

Reply:
Why is it not deterministic?

Reply:
The thread pool execute multiple tasks concurrently, the execution order of concurrent runnable instructions is not deterministic.

Reply:
> The thread pool execute multiple tasks concurrently, the execution order of concurrent runnable instructions is not deterministic.

That's indeed the purpose of testing it. It should still work correctly. Without concurrency, the unit tests will be trivial.

Reply:
That's why it's not necessary to test concurrency in unit tests. It won't deterministically catch a bug.

Reply:
Alright, deleted those concurrency tests.